feat(pose_instability_detector): change validation algorithm

TaikiYamada4 opened the pull request:
## Description

### Change of instability detection

Currently, `pose_instability_detector` compares the EKF pose and dead reckoning (DR) pose with a constant threshold defined in the `pose_instability_detector.param.yaml` file.

However, since the instability definition should change by the timer period and the vehicle twist, I made a framework for the `pose_instability_detector` to calculate the threshold of instability by itself, using the characteristics of the localization component. Note that this PR only puts timer period on consideration.

Hence, the `pose_instability_detector.param.yaml` drastically changed so that the parameters for calculation has been required but not the threshold itself.

Read the README.md for more information.

![Screenshot from 2024-06-03 14-37-49](https://github.com/autowarefoundation/autoware.universe/assets/129915538/d009f463-b2bf-4184-8d24-9fb3c4f0da56)

### Change of DR algorithm

Since there was a lack of time span for twist integration, I revised the algorithm so that the twist integration is fully done from the previous EKF pose to the current EKF pose.

## Related links

The following link shows the test results using actual driving data (For TIER IV people only)
[TIER IV INTERNAL LINK](https://tier4.atlassian.net/l/cp/CEn7007F)

## Tests performed

I have tested this module with several driving data in TIER IV, and check that the `/diagnostics` seems to output proper values. (For TIER IV people, see the link above.)
I also check that the `rqt_runtime_monitor` works well.

Currently, `pose_instability_detector` have output WARNs because of "large twist variance" and "angular twist error due to wrong gyro bias settings" even in normal environments but they appear very rare and only when the angular twist has a large perturbation.

## Notes for reviewers

<!-- Write additional information if necessary. It should be written if there are related PRs that should be merged at the same time. -->

## Interface changes

<!-- Describe any changed interfaces, such as topics, services, or parameters. -->

### ROS Topic Changes

None

### ROS Parameter Changes

Discarded all parameters, and added the following instead.

| Name                                      | Default Value | Description                                                     |
|-------------------------------------------|---------------|-----------------------------------------------------------------|
| timer_period                              | 0.5           | The period of timer_callback (sec).                             |
| heading_velocity_maximum                  | 16.667        | The maximum of heading velocity (m/s).                          |
| heading_velocity_scale_factor_tolerance   | 3.0           | The tolerance of heading velocity scale factor (%).             |
| angular_velocity_maximum                  | 0.523         | The maximum of angular velocity (rad/s).                        |
| angular_velocity_standard_deviation       | 0.00175       | The standard deviation of angular velocity (rad/s).             |
| angular_velocity_scale_factor_tolerance   | 0.2           | The tolerance of angular velocity scale factor (%).             |
| angular_velocity_bias_tolerance           | 0.00698       | The tolerance of angular velocity bias (rad/s).                 |
| pose_estimator_longitudinal_tolerance     | 0.11          | The tolerance of longitudinal position of pose estimator (m).   |
| pose_estimator_lateral_tolerance          | 0.11          | The tolerance of lateral position of pose estimator (m).        |
| pose_estimator_vertical_tolerance         | 0.11          | The tolerance of vertical position of pose estimator (m).       |
| pose_estimator_angular_tolerance          | 0.00631       | The tolerance of roll angle of pose estimator (rad).            |


## Effects on system behavior

This PR should not change the system behavior. However, since the current `pose_instability_detector` has optimistic threshold values and this PR makes it low, WARN diagnostics from this node may appear respectively more frequently (but yet with low possibility).

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

KYabuuchi commented on localization/pose_instability_detector/README.md:
```diff
@@ -1,20 +1,89 @@
+- $\omega_{\rm max} \cdots$ Maximum angular velocity [rad/s]
+- $\beta_\omega \cdots$ Scale factor tolerance for the maximum angular velocity [%]
+- $b \cdots$ Bias tolerance of the angular velocity [rad/s]
+- $\sigma_\omega \cdots$ Standard deviation of the angular velocity [rad/s]
```
I think the unit of  $\sigma_\omega$ is [ $rad/s^{3/2}$ ] .  Is it right?

Because the above equation has $3\sigma_\omega \sqrt{\Delta t} [rad/s]$

KYabuuchi commented on localization/pose_instability_detector/src/pose_instability_detector.cpp:
```diff
@@ -166,13 +170,258 @@ void PoseInstabilityDetector::callback_timer()
+    result_deque[result_deque.size() - 1].twist.twist.angular.z =
+      twist1.angular.z * ratio + twist0.angular.z * (1 - ratio);
+
+    result_deque[result_deque.size() - 1].header.stamp.sec =
+      static_cast<int32_t>(end_time.seconds());
+    result_deque[result_deque.size() - 1].header.stamp.nanosec =
+      static_cast<uint32_t>(end_time.nanoseconds() % 1000000000);
```
```suggestion
    result_deque[result_deque.size() - 1].header.stamp = end_time;
```

`rclcpp::Time` has a cast operator to `builtin_interfaces::msg::Time`, So you can directly substitude. 
https://github.com/ros2/rclcpp/blob/humble/rclcpp/src/rclcpp/time.cpp#L91-L104

KYabuuchi reviewed (commented):
I have not yet reviewed the detailed implementation, but I left some comments on what I noticed quickly.

KYabuuchi commented on localization/pose_instability_detector/src/pose_instability_detector.cpp:
```diff
@@ -166,13 +170,267 @@ void PoseInstabilityDetector::callback_timer()
+
+    const Twist twist = sliced_twist_deque[i - 1].twist.twist;
+
+    // quat to rpy
+    auto [ang_x, ang_y, ang_z] = quat_to_rpy(estimated_pose->orientation);
+
+    // average quaternion of two frames
+    tf2::Quaternion average_quat;
+    average_quat.setRPY(
+      ang_x + 0.5 * twist.angular.x * time_diff_sec, ang_y + 0.5 * twist.angular.y * time_diff_sec,
+      ang_z + 0.5 * twist.angular.z * time_diff_sec);
```
We should not convert quaternions to Euler angles, as it could lead to incorrect integral results when roll and pitch are non-zero.

TaikiYamada4 commented:
@KYabuuchi 
Thank you so much for your reviewing.
I believe I've fixed the dead reckoning algorithm when updating the orientation.

Anyways, I will revert this PR to draft since I have to evaluate this PR by myself since many changes have happened. Plus, I am to remove the consideration of the process noise in dead reckoning and I have to check that out too.

SakodaShintaro commented on localization/pose_instability_detector/CMakeLists.txt:
```diff
@@ -22,6 +24,14 @@ if(BUILD_TESTING)
   )
 endif()
 
+install(
+  DIRECTORY src/
+  DESTINATION include/pose_instability_detector
+  FILES_MATCHING PATTERN "*.hpp"
+)
+
+ament_export_include_directories(include)
+
```
If you want to be able to include this package from other modules, I think it is okay to explicitly create an include directory in the package.
https://autowarefoundation.github.io/autoware-documentation/main/contributing/coding-guidelines/ros-nodes/directory-structure/#include-and-src

TaikiYamada4 replied:
I followed the guidelines and revised the file structure. Thank you!
https://github.com/autowarefoundation/autoware.universe/pull/7042/commits/a462cf3340a311260f69a8cb954a45367b31a4e4 https://github.com/autowarefoundation/autoware.universe/pull/7042/commits/ff5218c735a3834cd5d0d575dc24f50199e43326

SakodaShintaro replied:
Oh, I remember we need to create an autoware directory. It is a good.

TaikiYamada4 commented on localization/pose_instability_detector/README.md:
```diff
@@ -34,4 +125,44 @@ The following diagram provides an overview of what the timeline of this process
+| Dead reckoning pose of corner C | $\left(1-\frac{\beta_v}{100}\right) v_{\rm max}$ | $\left(1-\frac{\beta_\omega}{100}\right) \omega_{\rm max} - b$ |
+| Dead reckoning pose of corner D | $\left(1+\frac{\beta_v}{100}\right) v_{\rm max}$ | $\left(1-\frac{\beta_\omega}{100}\right) \omega_{\rm max} - b$ |
+
+Given a heading velocity $v$ and $\omega$, the 2D theoritical variation seen from the previous pose is calculated as follows:
```
Fixed in https://github.com/autowarefoundation/autoware.universe/pull/7042/commits/3cc38342b511e0b0666c8713637312dc81de1027

TaikiYamada4 commented on localization/pose_instability_detector/package.xml:
```diff
@@ -33,5 +33,6 @@
 
   <export>
     <build_type>ament_cmake</build_type>
+    <include>include</include>
```
It's a left over by mistake. Thank you for pointing it out,
https://github.com/autowarefoundation/autoware.universe/pull/7042/commits/a772b1592a493088bdbee9c2a6c0e53a61bb33cb

TaikiYamada4 commented:
I've changed `calculate_threshold` method and `dead_reckon` method to public so that it can be used by other modules.
With this change, I also changed `calculate_threshold` to output the threshold values explicitly w/o changing the private variables.

KYabuuchi reviewed (approved):
Looks Good To Me :ok_man: 
Thank you for patiently responding to my many requests.
